### DISCUSSION TITLE
Support async I/O

### DISCUSSION THREAD

**Discussion by chenxi8611** [duckdb/duckdb#3560] [Ideas]
I would like to implement fully async I/O. I know it is big feature and probably touches a lot of places in code, but I think it is great feature. Adding async I/O would be a huge win for many of the clients connecting to DuckDB, allowing better persistency, async filesystem APIs, and tremendous improvements for any wasm browser-based implementations.

@Mytherin  I would love to work on this feature, and if you'd be open to this, I'd appreciate any suggestions on the main components in the codebase where I could begin implementing this.

**Comment by hannes** [comment]
This probably should start off as a discussion, not an issue.

**Comment by Mytherin** [comment]
I certainly agree that async I/O is a very desirable feature - but this feature would touch on so much of the core code that I think it is infeasible for this to be done by an external developer. This needs to be implemented by the core team. If this is important to your business we are happy to prioritize features, feel free to [get in touch](https://duckdblabs.com/contact/).

**Reply by foTok** [reply]
Async IO is pretty important. Now, checkpoint would lock all clients causing notable lag.

**Comment by danthegoodman1** [comment]
This would be very desirable for us as well. Being able to pull multiple parquet files from s3 as quickly as possible would be sublime! Especially if there was a way to query and save to disk (rather than memory cache, for lots of large files)

**Comment by davetapley** [comment]
Just found this, but Google also returned https://github.com/kouta-kun/aioduckdb

**Reply by Mause** [reply]
That project wraps the sync duckdb API with an async wrapper that does the sync work in threads, so it's subject to duckdb's thread limitations around read/write

**Reply by NickCrews** [reply]
Thanks for the explanation @Mause! Can you correct my understanding here if I'm wrong, I'd really appreciate it: Multithreaded could be faster than single threaded though, because when you enter the internals of a duckdb routine, the GIL gets released, so the interpreter is able to execute other python code as duckdb is churning away. But if you have multiple threads all doing a `con.execute()`, they all are going to happen sequentially, so the processing inside the database could still be the bottleneck.

**Reply by skinkie** [reply]
I have noticed that aioduckdb has a significant currency impact. When using the patterns from https://duckdb.org/docs/guides/python/multiple_threads.html the additional benefit is, the significantly less memory usage.

**Comment by david542542** [comment]
Any update on this item? Based on the duckdb community, does it seem like it'd ever get the necessary attention/priority it needs?

**Comment by paultiq** [comment]
For what it's worth, I'm not a big fan of duckdb extending the python lib with an asyncio API for a few reasons:
- Async support for most libs are largely syntactic sugar over the existing synchronous API. That feels cleaner as a separate overlay library rather than complicating the core codebase. This separation keeps maintenance easier and avoids mixing two execution models in one place.
-  As discussed in  [virtual threads discuss thread](https://discuss.python.org/t/add-virtual-threads-to-python/91403) maintainers often end up shouldering a lot of extra complexity when multiple concurrency models are supported. That can lead to bugs, divergence between sync/async paths, and slower evolution of the underlying library. This is further discussed in [what color is your function?](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).
- Python is adding other capabilities that can address the same core requirements: [subinterpreters / interpreterpoolexecutors](https://docs.python.org/3.14/whatsnew/3.14.html#pep-734-multiple-interpreters-in-the-stdlib) and [free-threading](https://docs.python.org/3.14/whatsnew/3.14.html#pep-779-free-threaded-python-is-officially-supported) ([discussion](https://github.com/duckdb/duckdb/discussions/18214)). 

I think free threading and/or subinterpreter support would be far more impactful and strategic than asyncio by itself.

**Reply by skinkie** [reply]
DuckDB (read-write) would be currently limited to a single _process_, regardless of binding. But from my perspective there is far bigger issue with DuckDB in is current form. Any ORDER BY query of a full table that yields resuls "caches" the **entire result**. Hence it does not matter that the client steps through the result, it process explodes in your face. The workaround? Not a MATERIALIZED VIEW - this is not supported in DuckDB - but first creating this intermediate table on disk, and then run a new query over it, and then dropping that table. Imagine this scenario with sub queries, transactions, etc.

**Reply by paultiq** [reply]
Not sure I follow - this seems unrelated to the topic of asyncio, or am I missing something?

**Reply by skinkie** [reply]
I thought _subinterpreters_ meant those would be runinng in new processes, not threads, don't know if this is valid or invalid. But anything that would use forking won't work in read-write mode with DuckDB as it is now. This would require a new process that does all the database handling and IPC from there with the other threads.

My point being: once you grow beyond synchronious code, all assumptions that DuckDB makes regarding memory availability are off. The user cannot naively do any query with an ORDER BY, the second concurrent query beyond that is more than main memory will bring the system to a grinding halt. Been there, done that.

**Reply by paultiq** [reply]
Subinterpreters and InterpreterPoolExecutors (subinterpreters in threads) do run in the same process, not forked, but yes - it'd work similarly to running the database in a separate process... except there's potential to share, say, arrow tables across the interpreters without copying. 

There's a longer thread here about how this might be done, and what other work needs to be done in various other projects. 

Regarding concurrency large memory queries - for sure - if each query is taking significant unbounded memory, multiple queries will run into memory contention... although I've had this issue with many databases. To the specific issue with "cannot naively do any query with an ORDER BY"... I think this is better taken in another discussion or GitHub issue.

**Comment by dentiny** [comment]
I did some POC work based libcurl multi-curl interface in my extension: https://duckdb.org/community_extensions/extensions/curl_httpfs.html

The core part lives in
- submit IO request and bind easy curl handle to multi-curl handle: https://github.com/dentiny/duckdb-curl-filesystem/blob/1928db20b7b53830abbb45bf010595845eedb4d8/src/multi_curl_manager.cpp#L374-L392
- use epoll / kqueue as polling engine for event-driven notification: https://github.com/dentiny/duckdb-curl-filesystem/blob/1928db20b7b53830abbb45bf010595845eedb4d8/src/multi_curl_manager.cpp#L308-L318

**Comment by dhirschfeld** [comment]
If async support is being added to the Python client, it would be great if `anyio` could be used so that users of other async frameworks such as `trio` can also benefit.

AnyIO implements modern, structured concurrency interfaces (as pioneered in Trio) but runs on top of either the `asyncio` or `trio` event-loops. Many popular projects such as FastAPI are now built on top of AnyIO to allow users their choice of async frameworks.

**Comment by dragonscraper** [comment]
any update on this ?

**Reply by Tishj** [reply]
https://github.com/duckdb/duckdb/pulls?q=is%3Apr+async+is%3Aclosed+sort%3Aupdated-desc

**Comment by david542542** [comment]
Cool, would this be in the next DuckDB release?

**Comment by szarnyasg** [comment]
Hello everyone, this was implemented and will be part of DuckDB v2.0.
See the blog post for more details: https://duckdb.org/2026/07/31/asynchronous-io

**Reply by skinkie** [reply]
With respect. This is not resolved. There is no explict memory budget able to be set. No mention of caching client interfaces, no line of async python code as (Python) example.

**Reply by Mytherin** [reply]
The Python client lives [here](https://github.com/duckdb/duckdb-python) and is not part of core DuckDB. If you want to make a (separate) feature request to support async I/O in the Python client then that belongs there, it has no place here and is not related to async I/O support in core DuckDB.

**Reply by davetapley** [reply]
- https://github.com/duckdb/duckdb-python/discussions/628
